File: parse/parser_test.go
package parse_test

import (
	"fmt"
	"path"
	"strings"
	"testing"

	"github.com/jabolopes/bapel/parse"
	"github.com/jabolopes/bapel/tests"
)

func TestParser(t *testing.T) {
	t.Parallel()

	matches, err := tests.Glob("testdata/in/*.in")
	if err != nil {
		t.Fatal(err)
	}

	for _, inFile := range matches {
		t.Run(inFile, func(t *testing.T) {
			t.Parallel()

			wantFile := strings.Replace(parse.ReplaceExtension(inFile, ".bpl"), "/in/", "/parsed/", 1)
			wantErr := strings.HasPrefix(path.Base(inFile), "bad_")

			module, err := parse.ParseSourceFile(inFile)
			if !wantErr && err != nil {
				t.Fatal(err)
			} else if wantErr && err == nil {
				t.Fatal("expected parse error but it succeeded")
			}

			var got string
			if wantErr {
				got = fmt.Sprintf("%s\n", err.Error())
			} else {
				got = fmt.Sprintf("%+s\n", module)
			}

			diff, err := tests.DiffOutRegen(got, wantFile)
			if err != nil {
				t.Fatal(err)
			}
			if len(diff) > 0 {
				t.Errorf("diff = %s", diff)
			}
		})
	}
}

File: tests/parser_test.cc
#include "tests/test_util.h"
#include "cpp_parser/parser.h"

#include <string>

TEST(ParserTest, GoldenFiles) {
  auto matches = tests::glob("tests/testdata/parse/in/*.in");
  ASSERT_FALSE(matches.empty());

  for (const auto& inFile : matches) {
    ctx.run(inFile, [&](tests::TestContext& sub_ctx) {
      auto directives = tests::TestDirectives::parse_from_file(inFile);
      if (!directives.should_run_stage("parse")) return;

      std::string wantFile = tests::replace_string(tests::replace_extension(inFile, ".bpl"), "/in/", "/parsed/");
      auto res = parser::parse_source_file_from_file(inFile);

      if (directives.expects_error("parse")) {
        if (res.ok) {
          sub_ctx.add_error("expected parse error for " + inFile + " but parser succeeded", __FILE__, __LINE__);
          return;
        }
      } else {
        if (!res.ok) {
          sub_ctx.add_error("parser failed on " + inFile + ":\n" + res.error(), __FILE__, __LINE__);
          return;
        }
      }

      std::string got = res.ok ? res.value.to_string(true /* with_pos */) : res.error();
      // Ensure trailing newline
      if (got.empty() || got.back() != '\n') {
        got += "\n";
      }

      std::string diff_str;
      if (!tests::diff_out_regen(got, wantFile, diff_str)) {
        sub_ctx.add_error("in test " + inFile + ":\n" + diff_str, __FILE__, __LINE__);
      }
    });
  }
}

